tempesta_fw/http_sess.c

```diff
@@ -0,0 +1,731 @@
+			  value->len, sizeof(StickyVal) * 2);
+		return TFW_BLOCK;
+	}
+
+	TFW_STR_FOR_EACH_CHUNK(c, value, end) {
+		for (p = c->ptr; p < (unsigned char *)c->ptr + c->len; ++p) {
+			if (i++ == sizeof(sv->ts) * 2)
+				goto ts_finished;
+			sv->ts = (sv->ts << 4) + hex_to_bin(*p);
+		}
+	}
+ts_finished:
+
+	if (__sticky_calc(req, sv))
+		return TFW_BLOCK;
+	for (i = 0, hi = 1; (c) < end; ++(c)) {
+		for ( ; p < (unsigned char *)c->ptr + c->len; ++p) {
+			b = hi ? hex_asc_hi(sv->hmac[i])
+			       : hex_asc_lo(sv->hmac[i]);
+			if (b != *p) {
+				int n = sizeof(sv->hmac) * 2;
+				char buf[n];
+				bin2hex(buf, sv->hmac, sizeof(sv->hmac));
+				sess_warn("bad sticky cookie value",
+					  addr, ": %c(pos=%d),"
+					  " ts=%#lx orig_hmac=[%.*s]\n",
+					  *p, i, sv->ts, n, buf);
+				return TFW_BLOCK;
+			}
+			hi = !hi;
+			i += hi;
+		}
+	}
+	BUG_ON(i != STICKY_KEY_MAXLEN);
+
+	/* Sticky cookie is found and verified, now we can set the flag. */
+	req->flags |= TFW_HTTP_HAS_STICKY_COOKIE;
+
+	return TFW_PASS;
+}
+
+/*
+ * Process Tempesta sticky cookie in an HTTP request.
+ */
+static int
+tfw_http_sticky_req_process(TfwHttpReq *req, StickyVal *sv)
+{
+	int r;
+	TfwStr cookie_val = {};
+
+	/*
+	 * See if the Tempesta sticky cookie is present in the request,
+	 * and act depending on the result.
+	 */
+	r = tfw_http_sticky_get(req, &cookie_val);
+	if (r < 0)
+		return r;
+	if (r == 0)
+		return tfw_http_sticky_notfound(req);
+	if (r == 1) {
+		/*
+		 * Verify sticky cookie value: if it's wrong, then this can be
+		 * an attack as well as changed Tempesta or whatever else.
+		 * The first case must be properly handled by Frang limit
+		 * (TODO #598), your ever can limit number of invalid sticky
+		 * cookie tries to 1. While we just send normal 302 redirect to
+		 * keep user experience intact.
+		 */
+		if (tfw_http_sticky_verify(req, &cookie_val, sv))
+			return tfw_http_sticky_send_302(req, sv) ? : 1;
+		return 0;
+	}
+	TFW_WARN("Multiple Tempesta sticky cookies found: %d\n", r);
+
+	return -1;
+}
+
+/*
+ * Add Tempesta sticky cookie to an HTTP response if needed.
+ */
+int
+tfw_http_sess_resp_process(TfwHttpResp *resp, TfwHttpReq *req)
+{
+	if (!tfw_cfg_sticky.enabled)
+		return 0;
+	BUG_ON(!req->sess);
+
+	/*
+	 * RFC 6265 4.1.1 and 4.1.2 says that we should not set session cookie
+	 * if it's not necessary. Since client didn't send up the cookie and
+	 * it seems that we don't enforce them, we can just set the cookie in
+	 * each response forwarded to the client.
+	 */
+	if (req->flags & TFW_HTTP_HAS_STICKY_COOKIE)
+		return 0;
+	return tfw_http_sticky_add(resp, req);
+}
+
+void
+tfw_http_sess_put(TfwHttpSess *sess)
+{
+	if (atomic_dec_and_test(&sess->users))
+		/*
+		 * Use counter reached 0, so session already expired and evicted
+		 * from the hash table.
+		 */
+		kmem_cache_free(sess_cache, sess);
+}
+
+/**
+ * Obtains appropriate HTTP session for the request based on Sticky cookies.
+ */
+int
+tfw_http_sess_obtain(TfwHttpReq *req)
```

Comment:
`obtain` is because we search and return or create a new session, just like `tfw_client_obtain()`. I wished to express the allocation/lookup logic rather than return value.